examples/perf/ipc.py

```diff
@@ -0,0 +1,186 @@
+    } else {
+        data.insert(&inst_k, &inst_start);
+    }
+
+    kptr = data.lookup(&time);
+    if (kptr) {
+        data.update(&time, &time_start);
+    } else {
+        data.insert(&time, &time_start);
+    }
+}
+
+void trace_end(struct pt_regs* ctx) {
+    u32 clk_k = 0;
+    u32 inst_k = 1;
+    u32 time = 2;
+
+    int cpu = bpf_get_smp_processor_id();
+    /*
+    perf_read may return negative values for errors.
+    If cpu id is greater than BPF_PERF_ARRAY size,
+    counters values will be very large negative number.
+    Modify MAX_CPUS to the current number of CPUs.
+    NOTE: Use bpf_perf_event_value is recommended over
+    bpf_perf_event_read or map.perf_read() due to
+    issues in ABI. map.perf_read_value() need to be
+    implemented in future.
+    */
+    u64 clk_end = clk.perf_read(cpu);
+    u64 inst_end = inst.perf_read(cpu);
+    u64 time_end = bpf_ktime_get_ns();
+    
+    struct perf_delta perf_data = {} ;
+    bool submit = true;
+    u64* kptr = NULL;
+    kptr = data.lookup(&clk_k);
+    if (kptr) {
+        perf_data.clk_delta = clk_end - *kptr;
+    }
+    
+    kptr = data.lookup(&inst_k);
+    if (kptr) {
+        perf_data.inst_delta = inst_end - *kptr;
+    } else {
+        submit = false;
+    }
+
+    kptr = data.lookup(&time);
+    if (kptr) {
+        perf_data.time_delta = time_end - *kptr;
+    } else {
+        submit = false;
+    }
+
+    if (submit) {
+        output.perf_submit(ctx, &perf_data, sizeof(struct perf_delta));
+    }
+}
+"""
+
+usage='Usage: ipc.py [options]\nexample ./ipc.py -l c -s strlen'
+parser = OptionParser(usage)
+parser.add_option('-l', '--lib', dest='lib_name', help='lib name containing symbol to trace, e.g. c for libc', type=str)
+parser.add_option('-s', '--sym', dest='sym', help='symbol to trace', type=str)
+
+(options, args) = parser.parse_args()
+if (not options.lib_name or not options.sym):
+    parser.print_help()
+    exit()
+
+b = BPF(text=code)
+
+# Attach Probes at start and end of the trace function
+# NOTE: When attaching to a function for tracing, during runtime relocation
+# stage by linker, function will be called once to return a different function
+# address, which will be called by the process. e.g. in case of strlen
+# after relocation stage, __strlen_sse2 is called instread of strlen.
+# NOTE: There will be a context switch from userspace to kernel space,
+# on caputring counters on USDT probes, so actual IPC might be slightly different.
+# This example is to give a reference on how to use perf events with tracing.
+b.attach_uprobe(name=options.lib_name, sym=options.sym, fn_name="trace_start")
+b.attach_uretprobe(name=options.lib_name, sym=options.sym, fn_name="trace_end")
+
+def print_data(cpu, data, size):
+    e = b["output"].event(data)
+    try:
+        print("%-8d %-8d %-8.2f %-8s %d" % (e.clk_delta, e.inst_delta, 
+            1.0* e.inst_delta/e.clk_delta, str(round(e.time_delta * 1e-3, 2)) + ' us', cpu))
+    except Exception as e:
```

Comment:
If the intent is to stop on `<ctrl+c>`, let's use `KeyError`

Reply:
Update: just saw you are handling this below. What kind of exception do you expect this code to raise?

Reply:
I was expecting zero division error, but since i now check for entries in map lookup and then only submit it to output event, the exception is not required now.